### PR TITLE
Change layernorms to two pass algorithm

### LINKED ISSUE #2238: [BUG] Nan value on GPU LayerNorm, valid value on CPU
**Describe the bug**
I run the same code with GPU and CPU and got vastly different results. One which is entire rows of NaN, the other valid values respectively. I isolated this layer with input, output, model weights, and bias from my model.

**To Reproduce**

Include code snippet
```
import mlx.core as mx
import mlx.nn as nn
import numpy as np
import json

# Set default device to GPU
mx.set_default_device(mx.gpu) # switch between mx.cpu and mx.gpu to observe the result

# Input vector (same as before)
x = mx.array([
    1449.570556640625,
    1448.8741455078125,
    1450.650390625,
    1449.2633056640625,
    1449.7763671875,
    1449.317138671875,
    1449.3297119140625,
    1449.077880859375,
    1449.2181396484375,
    1449.4329833984375,
    1449.1605224609375,
    1449.413330078125,
    1449.7933349609375,
    1450.066650390625,
    1449.7762451171875,
    1450.364990234375
])

# Load parameters
weight = mx.array([
        0.9819900989532471,
        0.8759878873825073,
        1.269553780555725,
        1.0803718566894531,
        0.8333801031112671,
        1.151047945022583,
        1.4396520853042603,
        0.9811643958091736,
        1.309312343597412,
        1.0645242929458618,
        1.4031972885131836,
        1.079842209815979,
        0.8111729621887207,
        1.4355424642562866,
        1.0123003721237183,
        1.261611819267273
    ])
bias = mx.array([
        -0.05786939337849617,
        0.00637853005900979,
        -0.029780251905322075,
        -0.01015615276992321,
        0.00462309829890728,
        -0.009934238158166409,
        0.0014344339724630117,
        -0.008547453209757805,
        0.011568562127649784,
        -0.004812314640730619,
        -0.0009176022140309215,
        -0.006406325846910477,
        0.015759265050292015,
        -0.02106950618326664,
        0.03796328231692314,
        0.04465882107615471
    ])

# Create LayerNorm
layernorm = nn.LayerNorm(16)
layernorm.weight = weight
layernorm.bias = bias

# Run LayerNorm
out = layernorm(x)

# Move result to numpy for inspection
out = out.tolist()
print("out:", out)
print("any nan:", np.any(np.isnan(out)))
```

**Expected behavior**
On CPU:
```
out: [0.7760803699493408, -192.16372680664062, 434.5673522949219, -104.06298065185547, 54.951255798339844, -91.27486419677734, -108.42243194580078, -152.0385284423828, -144.7915802001953, -45.412349700927734, -180.7537078857422, -52.77846908569336, 57.85071563720703, 226.40399169921875, 66.74213409423828, 318.0606689453125]
any nan: False
```
On GPU:
```
out: [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]
any nan: True
```

**Desktop (please complete the following information):**
 - OS Version: MacOS 15.4
 - Version 0.25.2 (latest mlx version)

**Additional context**
None

### PR DESCRIPTION
**Pull request by angeloskath**
This closes #2238. It was a bit tricky to get good performance in the end with the 2-pass algorithm. The gradients are generally slightly slower, the forward pass is same speed or a bit faster for small sizes and slightly slower for larger sizes. There are some big regressions of about 30% for some sizes unfortunately.

There may be ways to speed it up more if we need it.

Transformer training
--------------------

Before
```
Iter 10: Train loss 8.775, It/sec 1.580, Peak memory 15.703 (GB)
Iter 20: Train loss 8.167, It/sec 1.699, Peak memory 15.703 (GB)
Iter 30: Train loss 7.948, It/sec 1.701, Peak memory 15.703 (GB)
Iter 40: Train loss 7.792, It/sec 1.700, Peak memory 15.703 (GB)
```

After
```
Iter 10: Train loss 8.927, It/sec 1.581, Peak memory 15.703 (GB)
Iter 20: Train loss 8.265, It/sec 1.701, Peak memory 15.703 (GB)
Iter 30: Train loss 7.965, It/sec 1.702, Peak memory 15.703 (GB)
Iter 40: Train loss 7.840, It/sec 1.700, Peak memory 15.703 (GB)
```

Starcoder inference
--------------------

```
Before: 106.2
After: 106.7
```

Microbenchmark
-----------------

| Dtype | Size | Operation | Before | After |
| --- | --- | --- | --- | --- |
|float32 | 1024 | Forward | 2.07283      | 2.20885   |
|float32 | 1024 | Backward | 10.67157    | 10.73596  |
|float32 | 1024 | Backward X | 6.61871   | 6.61479   |
|float32 | 2048 | Forward | 3.77895      | 4.27993   |
|float32 | 2048 | Backward | 21.74962    | 21.80744  |
|float32 | 2048 | Backward X | 13.65268  | 13.66181  |
|float32 | 4096 | Forward | 10.64453     | 11.05335  |
|float32 | 4096 | Backward | 45.83887    | 45.81728  |
|float32 | 4096 | Backward X | 28.43235  | 28.34598  |
|float32 | 8192 | Forward | 23.90351     | 25.34620  |
|float32 | 8192 | Backward | 91.34822    | 92.21648  |
|float32 | 8192 | Backward X | 56.97122  | 59.47005  |
|float32 | 9216 | Forward | 26.88259     | 29.01165  |
|float32 | 9216 | Backward | 103.22656   | 104.8307  |
|float32 | 9216 | Backward X | 64.54061  | 66.03597  |
|float16 | 1024 | Forward | 1.90232      | 1.51465   |
|float16 | 1024 | Backward | 5.66858     | 5.64698   |
|float16 | 1024 | Backward X | 3.58427   | 3.31151   |
|float16 | 2048 | Forward | 3.70990      | 2.93282   |
|float16 | 2048 | Backward | 10.89499    | 10.80363  |
|float16 | 2048 | Backward X | 7.29720   | 6.73312   |
|float16 | 4096 | Forward | 6.34758      | 5.57065   |
|float16 | 4096 | Backward | 22.16688    | 21.84296  |
|float16 | 4096 | Backward X | 15.44423  | 14.58665  |
|float16 | 8192 | Forward | 11.29829     | 12.42783  |
|float16 | 8192 | Backward | 45.06726    | 47.03724  |
|float16 | 8192 | Backward X | 28.92226  | 34.92275  |
|float16 | 9216 | Forward | 13.22993     | 14.96959  |
|float16 | 9216 | Backward | 50.50502    | 55.94813  |
|float16 | 9216 | Backward X | 32.66263  | 41.33292  |
|bfloat16 | 1024 | Forward | 2.36096     | 2.14249   |
|bfloat16 | 1024 | Backward | 5.89813    | 5.90397   |
|bfloat16 | 1024 | Backward X | 3.65954  | 3.44194   |
|bfloat16 | 2048 | Forward | 4.17660     | 3.82541   |
|bfloat16 | 2048 | Backward | 11.29723   | 11.15373  |
|bfloat16 | 2048 | Backward X | 7.47700  | 6.96718   |
|bfloat16 | 4096 | Forward | 8.28806     | 7.26178   |
|bfloat16 | 4096 | Backward | 22.79703   | 22.24647  |
|bfloat16 | 4096 | Backward X | 15.85581 | 14.82577  |
|bfloat16 | 8192 | Forward | 17.97696    | 22.10782  |
|bfloat16 | 8192 | Backward | 46.54080   | 50.24536  |
|bfloat16 | 8192 | Backward X | 30.51765 | 35.98841  |
|bfloat16 | 9216 | Forward | 20.29531    | 26.66911  |
|bfloat16 | 9216 | Backward | 51.71390   | 62.87096  |
|bfloat16 | 9216 | Backward X | 34.18174 | 46.11187  |

### CODE REVIEW

File: mlx/backend/metal/normalization.cpp [hunk 1]
```diff
@@ -255,12 +255,13 @@ void LayerNorm::eval_gpu(
   auto axis_size = static_cast<uint32_t>(x.shape().back());
   int n_rows = x.data_size() / axis_size;
 
-  const int simd_size = 32;
-  const int n_reads = RMS_N_READS;
-  const int looped_limit = RMS_LOOPED_LIMIT;
+  int simd_size = 32;
+  int n_reads = 8;
+  int looped_limit = 6656;
```
Comment (awni) [inline mlx/backend/metal/normalization.cpp:260]: It might be worth making the static assert below (L276) a runtime check. I've seen in the past the max threads per thread group vary per device.. so if it ends up being smaller than 832 right now it will silently fail (without metal validation).

But if you feel pretty comfortable with that.. I'm also ok with it.

**Review by awni — APPROVED** [review]
Awesome! I think it's very much worth it given the numerical stability and mostly lack of perf regression from this.

**Comment by zcbenz** [comment]
The softmax kernel also uses online normalization, does it need the same change? 

Edit:

To answer my own question, I think the answer is no, the variance could become negative when doing online normalization in layer_norm, while softmax does not seem to suffer from this.

**Comment by awni** [comment]
> To answer my own question, I think the answer is no, the variance could become negative when doing online normalization in layer_norm, while softmax does not seem to suffer from this.

That's right. This is entirely due to the fact that in floating point math when computing the variance:

```python
mean((x - mean(x))**2)
```

is more sable than:

```python
mean(x**2) - mean(x)**2
```

But unfortunately the first one is harder to compute efficiently because the outer reduction depends on the inner one so it requires two steps.

**Comment by angeloskath** [comment]
Actually the numerical instability doesn't even require negative variance. `x - mean(x)` has strictly smaller absolute value that `x` which means that when the mean is large we are subtracting two very large values which is inherently imprecise.

For example in the issue that started this, the variance wasn't negative it was simply 0 instead of 0.2 because there were many significant digits in `mean(x)**2` and `mean(x**2)`.
